sql: fix internal failure with temp tables during ALTER operations

spilchen opened the pull request:
Certain operations—such as ALTER TYPE or ALTER DATABASE ... ADD REGION—could trigger an assertion failure if temporary tables were present. The failure occurred in planner.forEachMutableTableInDatabase, which used GetAllDescriptors. That method excludes temp schemas, so schema resolution for a temp table would fail.

The fix replaces GetAllDescriptors with GetAll, which includes temp schemas and avoids the failure.

Note: a broader catalog API issue around accessing temp objects across sessions (see issue #97822) remains unresolved.

Fixes #97975

Epic: none
Release note (bug fix): Fixed an internal assertion failure that could occur during operations like ALTER TYPE or ALTER DATABASE ... ADD REGION when temporary tables were present.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145551)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145551)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145551#-OPCA42gDmfgGowp0wqZ:-OPCA42gDmfgGowp0wq_:bmewcti):*
this bug doesn't seem specific to multiregion DBs. I went through the repro steps in https://github.com/cockroachdb/cockroach/issues/97975 on a local single-node cluster and it also hangs. let's be sure to update the release note as well.

```
create type my_enum as enum ('value1', 'value2');
set experimental_enable_temp_tables = on;
create temporary table tmp_table (id int primary key);
insert into tmp_table values (1), (2), (3);
ALTER TYPE my_enum ADD VALUE 'value3';
-- it hangs
```
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_temp` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145551#-OPCAOv5-n2LLze8vLHg:-OPCAOv5-n2LLze8vLHh:b-yql6x3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4be3658168a97a00cce38224c954c98de3fbe30/pkg/ccl/logictestccl/testdata/logic_test/multi_region_temp#L1)):*
> ```
> # LogicTest: multiregion-9node-3region-3azs
> ```

as per above, since this isn't specific to multiregion databases, we could keep this in its own test file


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145551)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145551#-OPCA42gDmfgGowp0wqZ:-OPCX7WN7GqNvMmL1gkh:b6zrrkh):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this bug doesn't seem specific to multiregion DBs. I went through the repro steps in https://github.com/cockroachdb/cockroach/issues/97975 on a local single-node cluster and it also hangs. let's be sure to update the release note as well.

```
create type my_enum as enum ('value1', 'value2');
set experimental_enable_temp_tables = on;
create temporary table tmp_table (id int primary key);
insert into tmp_table values (1), (2), (3);
ALTER TYPE my_enum ADD VALUE 'value3';
-- it hangs
```
</blockquote></details>

Good catch.
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_temp` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145551#-OPCAOv5-n2LLze8vLHg:-OPCX-2A9IyM-N8UtexV:b-lbnyk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4be3658168a97a00cce38224c954c98de3fbe30/pkg/ccl/logictestccl/testdata/logic_test/multi_region_temp#L1)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

as per above, since this isn't specific to multiregion databases, we could keep this in its own test file
</blockquote></details>

Ok, I moved the tests into existing test files since it's not specific to multiregion.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145551)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`-- commits` line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145551#-OPILxWq1H1ZJ_a9R193:-OPILxWq1H1ZJ_a9R194:b-ga2pi8):*
super nit: we don't need to mention the internal detail about descriptors in the release note, since most users won't know what that means, but it would be good to mention that in the commit message. perhaps instead of "missing," which makes it sound like it was a surprise to not find them, we should say something about how temp schemas do not have physical descriptors.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: fix internal failure with temp tables during ALTER operations

Certain operations—such as ALTER TYPE or ALTER DATABASE ... ADD REGION
—could trigger an assertion failure if temporary tables were present.
The failure occurred in planner.forEachMutableTableInDatabase, which
used GetAllDescriptors.  That method excludes temp schemas, so schema
resolution for a temp table would fail.

The fix replaces GetAllDescriptors with GetAll, which includes temp
schemas and avoids the failure.

Note: a broader catalog API issue around accessing temp objects across
sessions (see issue #97822) remains unresolved.

Fixes #97975

Epic: none
Release note (bug fix): Fixed an internal assertion failure that could
occur during operations like ALTER TYPE or ALTER DATABASE ... ADD REGION
when temporary tables were present.

spilchen commented:
TFTR!

bors r+
